### PR TITLE
Use callable object for patching dataloaders

### LINKED ISSUE #968: Passing dataloader to trainer.fit() doesn't work with tpu (and maybe ddp)
## 🐛 Bug

Receive a 
```
AttributeError: Can't pickle local object 'Trainer.__set_fit_dataloaders.<locals>.patch_train_dataloader'
``` 
error when passing the dataloader directly to `trainer.fit(model, train_loader)`

### To Reproduce

Try to call `trainer.fit(model, train_loader)` in TPU mode.

(I suspect that anything that calls `mp.spawn` will cause this problem, so ddp probably will face this issue too.)

#### Code sample
```python
import os

import pytorch_lightning as pl
import torch
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision import transforms
from torchvision.datasets import MNIST

import torch_xla.core.xla_model as xm


class CoolSystem(pl.LightningModule):
    def __init__(self, use_tpu=False):
        super(CoolSystem, self).__init__()
        # not the best model...
        self.use_tpu = use_tpu
        self.l1 = torch.nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

    def training_step(self, batch, batch_idx):
        # REQUIRED
        x, y = batch
        y_hat = self.forward(x)
        loss = F.cross_entropy(y_hat, y)
        tensorboard_logs = {'train_loss': loss}
        return {'loss': loss, 'log': tensorboard_logs}

    def validation_step(self, batch, batch_idx):
        # OPTIONAL
        x, y = batch
        y_hat = self.forward(x)
        return {'val_loss': F.cross_entropy(y_hat, y)}

    def validation_end(self, outputs):
        # OPTIONAL
        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
        tensorboard_logs = {'val_loss': avg_loss}
        return {'avg_val_loss': avg_loss, 'log': tensorboard_logs}

    def test_step(self, batch, batch_idx):
        # OPTIONAL
        x, y = batch
        y_hat = self.forward(x)
        return {'test_loss': F.cross_entropy(y_hat, y)}

    def test_end(self, outputs):
        # OPTIONAL
        avg_loss = torch.stack([x['test_loss'] for x in outputs]).mean()
        tensorboard_logs = {'test_loss': avg_loss}
        return {'avg_test_loss': avg_loss, 'log': tensorboard_logs}

    def configure_optimizers(self):
        # REQUIRED
        # can return multiple optimizers and learning_rate schedulers
        # (LBFGS it is automatically supported, no need for closure function)
        return torch.optim.Adam(self.parameters(), lr=0.0004)


if __name__ == '__main__':
    from pytorch_lightning import Trainer

    model = CoolSystem(use_tpu=True)

    dataset = MNIST(os.getcwd(),
                        train=True,
                        download=True,
                        transform=transforms.ToTensor())

    sampler = torch.utils.data.distributed.DistributedSampler(
        dataset,
        num_replicas=xm.xrt_world_size(),
        rank=xm.get_ordinal(),
        shuffle=True)

    loader = DataLoader(dataset, sampler=sampler, batch_size=32)

    # most basic trainer, uses good defaults
    trainer = Trainer(num_tpu_cores=8)
    trainer.fit(model, loader)
```

### Expected behavior

Ideally, specifying the dataloaders as part of the LightningModule should work just the same as passing the dataloaders into `trainer.fit()`

### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```
 - Docker image: gcr.io/tpu-pytorch/xla:nightly
 - build steps:
     - pip install git+git://github.com/williamFalcon/pytorch-lightning.git@master --upgrade 

(I don't have access to the machine right now, so please forgive me on the specific version info temporarily)

### Proposed solution

The issue is [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/d856989120b078581f3f694fd7a1c036703f67a9/pytorch_lightning/trainer/trainer.py#L1001), trying to assign a local function to the model
```python
def __set_fit_dataloaders(self, model, train_dataloader, val_dataloaders, test_dataloaders):
        # when dataloader is passed via fit, patch the train_dataloader
        # functions to overwrite with these implementations
        if train_dataloader is not None:
            if not self.is_overriden('training_step', model):
                m = 'You called .fit() with a train_dataloader but did not define training_step()'
                raise MisconfigurationException(m)

            def patch_train_dataloader():
                return train_dataloader

            model.train_dataloader = patch_train_dataloader
```

Instead of using a closure or a local function, you could [use a callable](https://stackoverflow.com/questions/573569/python-serialize-lexical-closures) defined at the top-level.  This will be pickleable.

```python

class DataLoaderPatcher
    def __init__(self, loader):
        self.loader = loader
    def __call__():
        return self.loader

def __set_fit_dataloaders(self, model, train_dataloader, val_dataloaders, test_dataloaders):
        # when dataloader is passed via fit, patch the train_dataloader
        # functions to overwrite with these implementations
        if train_dataloader is not None:
            if not self.is_overriden('training_step', model):
                m = 'You called .fit() with a train_dataloader but did not define training_step()'
                raise MisconfigurationException(m)

            model.train_dataloader = DataLoaderPatcher(train_dataloader)
```

### PR DESCRIPTION
**Pull request by shoarora**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
    - not sure if there's anywhere I need to
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #968 .

Changes how we patch dataloaders when passed to `trainer.fit()` so that it's pickleable

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by shoarora** [comment]
@williamFalcon @Borda 

Do we need tests for this?
Not sure if this needs a CHANGELOG change either.  would appreciate review!

**Comment by shoarora** [comment]
Getting a very unhelpful error, will keep investigating

```
Exception                                 Traceback (most recent call last)
<ipython-input-15-bef80bcdc7b4> in <module>()
     16 # trainer = Trainer(num_tpu_cores=8)
     17 
---> 18 trainer.fit(model, loader)

2 frames
/usr/local/lib/python3.6/dist-packages/torch/multiprocessing/spawn.py in join(self, timeout)
    110                 raise Exception(
    111                     "process %d terminated with exit code %d" %
--> 112                     (error_index, exitcode)
    113                 )
    114 

Exception: process 0 terminated with exit code 1
```

**Comment by shoarora** [comment]
Update: now if you pass your dataloaders to `trainer.fit()`:
- non-slurm ddp works in a VM/linux context
- non-slurm ddp does not work in colab
    - the colab stack trace (my previous message) is unhelpful, and I'm having trouble figuring out how to get to the bottom of it
    - on the other hand, ddp in a colab context is an impractical scenario
- colab TPU worked regardless of this PR
- non-colab TPU works

Not sure what a change of this level warrants in terms of testing and documentation for pl.  Please let me know!

I added a gpu-ddp test for passing all the loaders to fit() and confirmed that this branch allows the test to pass

**Comment by williamFalcon** [comment]
this is awesome! will pr review in a few hours!
thank u

**Comment by williamFalcon** [comment]
i think ddp in colab needs the spawn flag or something

**Comment by shoarora** [comment]
> LGTM, just curious if simple `functools` would help too which does very similar wrapping
> 
> ```python
> from functools import partial
> 
> dataloader = partial(dataloader)
> ```

`partial` is specifically for wrapping functions/callables and freezing sets of arguments, so you couldn't call it on a dataloader:

```
>>> from functools import partial
>>> x = partial(loader)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: the first argument must be callable
```

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1189,6 +1180,21 @@ def test(self, model: Optional[LightningModule] = None):
...
 
+class _PatchDataLoader(object):
+    r'''
+    Callable object for patching dataloaders passed into trainer.fit().
+    Use this class to override model.*_dataloader() and be pickle-compatible.
+
+    Args:
+        dataloader: Dataloader object to return when called.
+    '''
+    def __init__(self, dataloader):
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]:
```suggestion
    def __init__(self, dataloader: DataLoader):
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1189,6 +1180,21 @@ def test(self, model: Optional[LightningModule] = None):
...
+        self.dataloader = dataloader
+
+    def __call__(self):
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]:
```suggestion
    def __call__(self) -> DataLoader:
```

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (shoarora) [reply pytorch_lightning/trainer/trainer.py]: For the val and test cases, could we receive a list of dataloaders?  the variable names are plural: `val_dataloaders` and `test_dataloaders`.  Are they lists of loaders by this point?

In that case it should be 
`-> Union[list, DataLoader]` correct?

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: true so rather `-> Union[List[DataLoader], DataLoader]`

**Pull request merged**